File: Trees/C_Cpp/delete_node_BST.c
#include <stdlib.h>
struct TreeNode {
    int val;
    struct TreeNode *left;
    struct TreeNode *right;
};

int min(struct TreeNode* root);

struct TreeNode* deleteNode(struct TreeNode* root, int key) {
    if (!root) {
        return NULL;
    }

    if (key > root->val){
        root->right = deleteNode(root->right, key);
    }
    else if (key < root->val){
        root->left = deleteNode(root->left, key);
    }
    else {
        if (!root->right) {
            return root->left;
        }
        else if (!root->left) {
            return root->right;
        }
        else {
            int mv = min(root->right);
            root->val = mv;
            root->right = deleteNode(root->right, mv);
        }
    }


    return root;
}


int min(struct TreeNode* root) {
    while (root && root->left) {
        root = root->left;
    }
    return root->val;
}

File: Trees/Python/delete_node_BST.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def minNode(self, root):
        cur = root
        while cur and cur.left:
            cur = cur.left
        return cur.val

    def deleteNode(self, root: Optional[TreeNode], key: int) -> Optional[TreeNode]:
        if not root:
            return None
        if key > root.val:
            root.right = self.deleteNode(root.right, key)
        elif key < root.val:
            root.left = self.deleteNode(root.left, key)
        else:
            if not root.right:
                return root.left
            elif not root.left:
                return root.right
            else:
                minVal = self.minNode(root.right)
                root.val = minVal
                root.right = self.deleteNode(root.right,minVal)
        return root


